lib/user_comm_screener.rb

```diff
@@ -0,0 +1,208 @@
+# frozen_string_literal: true
+
+# There are various ways within Discourse that a user can prevent
+# other users communicating with them. The purpose of this class is to
+# find which of the target users are ignoring, muting, or preventing
+# private messages from the acting user, so we can take alternative
+# action (such as raising an error or showing a helpful message) if so.
+#
+# Users may Mute another user (the actor), which will:
+#
+#   * Prevent PMs from the actor
+#   * Prevent notifications from the actor
+#
+# Users may Ignore another user (the actor), which will:
+#
+#   * Do everything that Mute does as well as suppressing content made by
+#     the actor (such as posts) from the UI
+#
+# Users may also either:
+#
+#   a) disallow all PMs sent to them or
+#   b) disallow PMs except from a certain allowlist of users
+#
+# A user may have this preference but have no Muted or Ignored users, which
+# necessitates the difference between methods in this class.
+#
+# An important note is that **all of these settings do not apply when the actor
+# is a staff member**. So admins and moderators can PM and notify anyone they please.
+class UserCommScreener
+  attr_reader :acting_user, :preferences
+
+  class UserCommPref
```

Comment:
Not sure if this and `UserCommPrefs` should be in the same file but at the same time don't really want anything else using them directly, I have this dilemma any time I want a private class really.